proposals/scripting-support.md

```diff
@@ -0,0 +1,434 @@
+
+For the command line usage the support for script parameters is needed. The simplest form is to assume that the script
+has access to the `args: Array<String>` property/parameter. More advanced is to support a declaration of the typed 
+parameters, e.g.:
+
+```
+@file:param("name", String::class)
+@file:param("num", Int::class)
+
+// this script could be called with args "-name=abc -num=42" 
+// and then in the body we can access parsed typed arguments
+
+println("$name ${num/6}") 
+```
+
+#### IDE support
+
+The IDE counterpart of the standalone scripting support should be able to extract execution environment properties from 
+the script itself and some explicit configuration data and create appropriate virtual project/module environment that 
+is able to resolve all symbols properly used in the script, including external libraries and scripts.
+
+#### Standalone REPL
+
+Standalone REPL is invoked by a dedicated host the same way as for standalone script but accepts user's input as script 
+lines. It means that the declarations made on the previous lines are accessible on the subsequent ones.
+
+### Project-level scripting
+
+Applications: test definition scripts (Spek), ktor routing scripts, project-level REPL, build scripts, type safe 
+config files, etc.
+ 
+Project-level scripts are executed by some dedicated scripting host usually embedded into the project build system. 
+From an IDE point of view, they are project-context dependent, but may not be part of the project sources. (In the same 
+sense as e.g. gradle build scripts source is not considered as a part of the project sources.)
+
+#### Discovery
+
+The IDE needs to be able to extract scripts environment configurations from the project settings.
+ 
+#### Project-level REPL
+ 
+A REPL that has access to the project's compiled classes. 
+
+## Proposal
+
+### Architecture
+
+The scripting support consists of the following components:
+- **Platform** - interface for compilation and IDE support 
+  - `(scriptSource, scriptDefinition, compilationConfig) -> compiledScript`
+  - `(scriptSource, scriptDefinition, compilationConfig) -> ideDataForRichEditing`
+  - predefined platforms based on the kotlin platforms: /JVM, /JS, /Native
+  - custom/customized implementation possible
+  - compiled scripts cashing should be implemented here
+  - the `compilationConfig` is defined by the platform, and may affect code generation; typically contains:
+    - additional dependencies
+    - *required* scripts
+    - expected parameters
+  - should not keep the state of the script execution
+- **Executor** - the component that receives compiled script instantiates it, and then executes it in a required 
+  environment, supplying any arguments that the script requires:
+  - `(compiledScript, compilationConfig, environment) -> scriptObject`
+  - `(scriptObject, compilationConfig, environment) -> Unit`
+  - uses `compilationConfig` and `environment` for "feeding" the script with the arguments
```

Comment:
If "feeding" is not a specially coined term, replace it with "passing arguments to the script"